### ISSUE TITLE
Problem validating certificates with ruby 2.5.3 and openssl 1.1.1a

### ISSUE THREAD

**Issue by Keruspe** [ruby/openssl#238]
I think this is the right place to report this. When I try to run this little reproducer:

```
require 'net/imap'

Net::IMAP::new('imap.gmail.com', :port => port=993, :ssl => true)
```

I get 

```
Traceback (most recent call last):
        5: from test-imap.rb:3:in `<main>'
        4: from test-imap.rb:3:in `new'
        3: from /usr/x86_64-pc-linux-gnu/lib/ruby/2.5/net/imap.rb:1092:in `initialize'
        2: from /usr/x86_64-pc-linux-gnu/lib/ruby/2.5/net/imap.rb:1531:in `start_tls_session'
        1: from /usr/x86_64-pc-linux-gnu/lib/ruby/2.5/net/protocol.rb:44:in `ssl_socket_connect'
/usr/x86_64-pc-linux-gnu/lib/ruby/2.5/net/protocol.rb:44:in `connect_nonblock': SSL_connect returned=1 errno=0 state=error: certificate verify failed (self signed certificate) (OpenSSL::SSL::SSLError)
```

No problem when building and running with openssl 1.0.2. `openssl s_client` from openssl 1.1.1 doesn't report any problem either, so it seems to be a bad interaction between ruby-openssl and openssl 1.1.1

**Comment by Keruspe** [comment]
Note that with openssl 1.0.2, ruby tries to open "/etc/ssl/certs/4a6481c9.0" which does exist, and with openssl 1.1.0 it tries to open "/etc/ssl/certs/3a77ab42.0" which does not. Looks like some problem in some hashing primitive maybe?

**Comment by Keruspe** [comment]
`openssl s_client -connect imap.gmail.com:993` using openssl 1.1.1 also uses "/etc/ssl/certs/4a6481c9.0"

**Comment by lindes** [comment]
Is this fixed in Ruby 2.6.3, with the incorporation of the above-mentioned pull request?

Because 2.6.3 fixes code similar to your initial reproducer, for me, though it's unclear to me whether this report would be fixed by that, or is more specific to particular openssl versions.

More on my [own solution](https://stackoverflow.com/a/56065615/313756), in case you and/or others find it helpful; short version:

1. Fixed for me in 2.6.3, but one can also:
2. use `ssl: {ssl_version: :TLSv1_2}` instead of `:ssl => true` as a workaround.

**Comment by Keruspe** [comment]
Looks like it was backported to 2.6.3, yes
